go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go

```diff
@@ -32,6 +49,888 @@ func (c *Controller) processNextANPWorkItem(wg *sync.WaitGroup) bool {
+			}
+			ports = append(ports, lsp)
+			anpSubject.podPorts.Insert(lsp.UUID)
+			podCache[pod.Name] = true
+		}
+	}
+	anpSubject.namespaces = namespaceCache
+
+	return ports, nil
+}
+
+// clearAdminNetworkPolicy will handle the logic for deleting all db objects related
+// to the provided anp which got deleted.
+// uses externalIDs to figure out ownership
+func (c *Controller) clearAdminNetworkPolicy(anpName string) error {
+	anp, loaded := c.anpCache[anpName]
+	if !loaded {
+		// there is no existing ANP configured with this name, nothing to clean
+		klog.Infof("ANP %s not found in cache, nothing to clear", anpName)
+		return nil
+	}
+
+	// clear NBDB objects for the given ANP (PG, ACLs on that PG, AddrSets used by the ACLs)
+	var err error
+	// remove PG for Subject (ACLs will get cleaned up automatically)
+	portGroupName, readableGroupName := getAdminNetworkPolicyPGName(anp.name)
+	// no need to batch this with address-set deletes since this itself will contain a bunch of ACLs that need to be deleted which is heavy enough.
+	err = libovsdbops.DeletePortGroups(c.nbClient, portGroupName)
+	if err != nil {
+		return fmt.Errorf("unable to delete PG %s for ANP %s: %w", readableGroupName, anp.name, err)
+	}
+	// remove address-sets that were created for the peers of each rule fpr the whole ANP
+	// do this after ACLs are gone so that there is no lingering references
+	err = c.clearASForPeers(anp.name, libovsdbops.AddressSetAdminNetworkPolicy)
+	if err != nil {
+		return fmt.Errorf("failed to delete address-sets for ANP %s/%d: %w", anp.name, anp.priority, err)
+	}
+	// we can delete the object from the cache now.
+	delete(c.anpPriorityMap, anp.priority)
+	delete(c.anpCache, anpName)
+
+	return nil
+}
+
+// clearASForPeers takes the externalID objectIDs and uses them to delete all the address-sets
+// that were owned by anpName
+func (c *Controller) clearASForPeers(anpName string, idType *libovsdbops.ObjectIDsType) error {
+	predicateIDs := libovsdbops.NewDbObjectIDs(idType, c.controllerName,
+		map[libovsdbops.ExternalIDKey]string{
+			libovsdbops.ObjectNameKey: anpName,
+		})
+	asPredicate := libovsdbops.GetPredicate[*nbdb.AddressSet](predicateIDs, nil)
+	if err := libovsdbops.DeleteAddressSetsWithPredicate(c.nbClient, asPredicate); err != nil {
+		return fmt.Errorf("failed to destroy address-set for ANP %s, err: %v", anpName, err)
+	}
+	return nil
+}
+
+// createNewANP takes the desired state of the anp and creates the corresponding objects in the NBDB
+func (c *Controller) createNewANP(desiredANPState *adminNetworkPolicyState, desiredACLs []*nbdb.ACL, desiredPorts []*nbdb.LogicalSwitchPort) error {
+	ops := []ovsdb.Operation{}
+	var err error
+	portGroupName, readableGroupName := getAdminNetworkPolicyPGName(desiredANPState.name)
+	pgExternalIDs := map[string]string{ANPExternalIDKey: readableGroupName}
+	for _, rule := range desiredANPState.ingressRules {
+		addrOps, err := rule.addrSet.AddIPsReturnOps(stringsToIPs(rule.podIPs.UnsortedList()))
```

Comment:
good question.. I can see why this is confusing..
buildACL -> createASForPeers -> creates the address-set and assigns it to the cache at the same time.. So when we reach this spot all address-sets are already created

Reply:
Right, this is tricky. It's happening in `convertANPRuleToACL` as a side effect. That calls to `createASForPeers` which  returns both the AS and the IPs but without setting the IPs in the AS, so I wonder why those two things need to be done in the same method at all.

Wouldn't it be clear to have a single entry point to build the desiredANPState that gives us a final state set as we need it to be instead of updating it as we go with distributed side effects around the code?

Reply:
yeah let me see if I can do a single entry for all the ops so that its more clean.

Reply:
ah I remember now why i split the `buildANPACLs` into its own thing - its because its also used in update logic so it was easier to handle..

Reply:
all right i split out the address set add from acl add.
made it independent
that also made me clear up storing the address-sets in the cache which is no longer really needed since we can use incides of address-sets to pull it out since we changed to using indices instead of priorities -> we know this part won't change.
resolving this comment.